### PR TITLE
Tracing private transactions feature

### PR DESCRIPTION
**Pull request by NickSneo**
## PR description
PR for adding priv_traceTransaction API

## Fixed Issue(s)
https://github.com/hyperledger/besu/issues/5280

### CODE REVIEW

**Comment by macfarla** [comment]
Hi @NickSneo - would like to see some unit and acceptance tests for this new method!

**Comment by NickSneo** [comment]
> Hi @NickSneo - would like to see some unit and acceptance tests for this new method!

Hey @macfarla , Sure. I am just working on it. Will push the new commits by the end of this week.

**Review by pinges — CHANGES_REQUESTED** [review]
I've noticed that there is a lot of code that is very similar to existing code (e.g. see PrivateTraceGenerator vs FlatTraceGenerator).
I think it would be good to try to have less code duplication.

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    /*
+     Updating the contract value
+    */
+    Hash transactionHash =
+        Hash.fromHexString(doTransaction(privacyGroupId, simpleStorageContract, 1));
+
+    final String result =
+        node.execute(privacyTransactions.privTraceTrasnaction(privacyGroupId, transactionHash));
+
+    System.out.println("privTransactionTrace = " + result);
```
[the remark this answers is not in the document]
Comment (NickSneo) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java]: Added more checks on the result

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    assertThat(result).isNotNull();
+  }
+
```
[the remark this answers is not in the document]
Comment (NickSneo) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java:139]: added checks for wrong PrivacyID, wrong node, and wrong Tx Hash

**Comment by NickSneo** [comment]
> I've noticed that there is a lot of code that is very similar to existing code (e.g. see PrivateTraceGenerator vs FlatTraceGenerator). I think it would be good to try to have less code duplication.

Hey @pinges ,

Thank you for reviewing the PR. I have resolved all the comments. For code duplication I have tried to have as less as possible, but since we want to make private and public code separate from each other, I have followed the similar approach as it is right now -> some code duplication is expected. Let me know if it is fine or I still need to change.

Also regarding the failing pipeline, It is not related to any of my change, should I try to fix that too in this PR?

**Comment by NickSneo** [comment]
Hey @pinges @macfarla ,
Please review this PR

> > I've noticed that there is a lot of code that is very similar to existing code (e.g. see PrivateTraceGenerator vs FlatTraceGenerator). I think it would be good to try to have less code duplication.
> 
> Hey @pinges ,
> 
> Thank you for reviewing the PR. I have resolved all the comments. For code duplication I have tried to have as less as possible, but since we want to make private and public code separate from each other, I have followed the similar approach as it is right now -> some code duplication is expected. Let me know if it is fine or I still need to change.
> 
> Also regarding the failing pipeline, It is not related to any of my change, should I try to fix that too in this PR?

File: plugin-api/src/main/java/org/hyperledger/besu/plugin/data/PrivateTransaction.java [hunk 3]
```diff
@@ -94,6 +94,17 @@ public interface PrivateTransaction {
    */
   Address getSender();
 
+  /**
+   * An unlimited size byte array specifying the EVM-code for the account // initialisation
```
Comment (macfarla) [inline plugin-api/src/main/java/org/hyperledger/besu/plugin/data/PrivateTransaction.java:98]:
```suggestion
   * An unlimited size byte array specifying the EVM-code for the account initialization
```

**Comment by macfarla** [comment]
> Hey @pinges @macfarla , Please review this PR
> 
> > > I've noticed that there is a lot of code that is very similar to existing code (e.g. see PrivateTraceGenerator vs FlatTraceGenerator). I think it would be good to try to have less code duplication.
> > 
> > 
> > Hey @pinges ,
> > Thank you for reviewing the PR. I have resolved all the comments. For code duplication I have tried to have as less as possible, but since we want to make private and public code separate from each other, I have followed the similar approach as it is right now -> some code duplication is expected. Let me know if it is fine or I still need to change.
> > Also regarding the failing pipeline, It is not related to any of my change, should I try to fix that too in this PR?

would it make sense to have an abstract class that FlatTrace can extend?

For the failing test - i have created an issue for it #7108 -  that is not caused by this PR - but a lot of those permissioning tests are flaky lately - so if you feel like fixing them separately that would be super!

**Comment by NickSneo** [comment]
> would it make sense to have an abstract class that FlatTrace can extend?
> 
> For the failing test - i have created an issue for it #7108 - that is not caused by this PR - but a lot of those permissioning tests are flaky lately - so if you feel like fixing them separately that would be super!

@macfarla  Sure, will create an abstract class that FlatTrace can extend. Also will look into the issue #7108

**Comment by NickSneo** [comment]
> would it make sense to have an abstract class that FlatTrace can extend?
> 
> For the failing test - i have created an issue for it #7108 - that is not caused by this PR - but a lot of those permissioning tests are flaky lately - so if you feel like fixing them separately that would be super!

Hey @macfarla ,
I tried two approaches for FlatTraceGenerator and PrivateTraceGenerator - 

1. Abstract Class - But the signatures are different, and I tried to have a common signature but then goes into a rabbit hole to change a lot of classes but still failed to resolve few issues like handling ExecutedPrivateTransaction and Transaction.
2. PrivateTraceGenerator inheriting FlatTraceGenerator - In this too failed to override common methods since singatures are different.

Right now not trying to modify logic for FlatTraceGenerator and PrivateTraceGenerator.

For FlatTrace and PrivateFlatTrace - I made the PrivateFlatTrace to inherit from FlatTrace with custom freshBuilder function.

I will push these changes, kindly let me know if above approaches sounds good?

Also, Can we get this merged soon, so I can pick other issues too. Thanks

**Comment by NickSneo** [comment]
Hey @macfarla @non-fungible-nelson ,

Can we get this PR merged soon? It is open from a long time, also I have spent a lot of time on it so it would be great to see this merged. Thanks!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/privacy/privateTracing/PrivateTraceGenerator.java [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.collect.Streams;
+import com.google.common.util.concurrent.Atomics;
+import org.apache.tuweni.bytes.Bytes;
+
+public class PrivateTraceGenerator {
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/privacy/privateTracing/PrivateTraceGenerator.java:52]: This class looks a lot like the normal TraceGenerator. Does it make sense for the PrivateTraceGenerator to extend the normal one?

**Review by pinges — COMMENTED** [review]
Please have a look at the comments.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/privacy/privateTracing/PrivateTraceGenerator.java (hunk 4 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/privacy/privateTracing/PrivateTraceGenerator.java:52]: Saw that you have put a lot of effort into this already. That should be fine.

**Review by pinges — COMMENTED** [review]
@NickSneo I think we are very close to get this merged. If my comments about the gas do not make sense just let me know :-)

**Comment by NickSneo** [comment]
> @NickSneo I think we are very close to get this merged. If my comments about the gas do not make sense just let me know :-)

Hey @pinges , Thanks for reviewing. Comments make sense, actually I got stuck in other Besu PR - EIP196 constantine bindings. I will resolve the comments and push the changes

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/processor/privateProcessor/PrivateBlockReplay.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                                          transactionProcessor,
+                                          dataGasPrice))
+                              .orElse(null))
+                  .toList();
+
+          return Optional.of(new PrivateBlockTrace(transactionTraces));
+        });
+  }
+
+  public <T> Optional<T> performActionWithBlock(final Hash blockHash, final BlockAction<T> action) {
```
[the remark this answers is not in the document]
Comment (NickSneo) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/processor/privateProcessor/PrivateBlockReplay.java:96]: removed

File: plugin-api/build.gradle [hunk 6]
```diff
@@ -70,7 +70,7 @@ Calculated : ${currentHash}
 tasks.register('checkAPIChanges', FileStateChecker) {
   description = "Checks that the API for the Plugin-API project does not change without deliberate thought"
   files = sourceSets.main.allJava.files
-  knownHash = 'Q6EK5By3BNKNa/JYqYjFw43VXWL0KVBUV3dGEQBjZ70='
+  knownHash = 'RXb/hQoaO7oodmrfIhQRkZvYaZ3j/9LFekHOMtmMKNE='
```
[the remark this answers is not in the document]
Comment (NickSneo) [reply plugin-api/build.gradle:73]: It is giving me error if I don't change the expected hash

**Comment by NickSneo** [comment]
Hey @pinges Sorry for the delay on this, I just pushed new commits to resolve comments. Please review.
Thanks!

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            false,
+            false,
+            restriction == UNRESTRICTED);
+
+    privacyCluster.start(node);
+    privacyCluster.start(wrongNode);
+  }
+
+  @Test
+  public void getTransactionTrace() throws JsonProcessingException {
```
Comment (pinges) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java:77]: Could you add a test verifying that in multi tenancy you can only access your own privacy groups.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivTraceTransaction.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  @Override
+  public String getName() {
+    return RpcMethod.PRIV_TRACE_TRANSACTION.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+
+    final Hash transactionHash = requestContext.getRequiredParameter(1, Hash.class);
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivTraceTransaction.java:71]: your not catching any exceptions, like illegal parameters ...

**Review by macfarla — COMMENTED** [review]
needs a changelog entry

**Review by macfarla — COMMENTED** [review]
the privacy ATs are not running by default in GHA, have you run the new ATs locally to ensure they pass?

**Review by pinges — COMMENTED** [review]
Couple of small things. Please add a test to check multi tenancy. Otherwise looking good!

**Comment by NickSneo** [comment]
> the privacy ATs are not running by default in GHA, have you run the new ATs locally to ensure they pass?

I ran the Privacy Acceptance tests - `./gradlew acceptance-tests:tests:acceptanceTestPrivacy`

195 tests completed, 161 failed, 2 skipped

All the tests failed due to - `org.testcontainers.containers.ContainerLaunchException: Container startup failed for image quorumengineering/tessera:22.10.0`

Seems issue with tessera. Running on Mac M2, openJDK 21. Any idea how can I fix the tessera thing or do I need to change any configuration or setting?

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java (hunk 7 above)
Comment (NickSneo) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java:77]: Hey @pinges, I looked into current acceptance tests (also in privacy/multitenancy tests) but I couldn't find a similar verification test for any priv RPC API. Can you point me to a similar test to which I can refer?

File: plugin-api/build.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (macfarla) [reply plugin-api/build.gradle]: there is a small change to one file in plugin-api so the hash does need updating

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java (hunk 7 above)
Comment (macfarla) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/privacy/PrivTraceTransactionAcceptanceTest.java:77]: I think actually having another look, the createWrongPrivacyGroup() method call covers this case

**Comment by macfarla** [comment]
> > the privacy ATs are not running by default in GHA, have you run the new ATs locally to ensure they pass?
> 
> I ran the Privacy Acceptance tests - `./gradlew acceptance-tests:tests:acceptanceTestPrivacy`
> 
> 195 tests completed, 161 failed, 2 skipped
> 
> All the tests failed due to - `org.testcontainers.containers.ContainerLaunchException: Container startup failed for image quorumengineering/tessera:22.10.0`
> 
> Seems issue with tessera. Running on Mac M2, openJDK 21. Any idea how can I fix the tessera thing or do I need to change any configuration or setting?

argh the tessera container tests - we tried so hard to refactor these to make them less brittle but haven't looked at them for a while. they try to pull in tessera as a dependency via a docker image, idk why it's not working - this is a problem with having tests that we aren't running. I consider it non-blocking for this PR

**Comment by NickSneo** [comment]
@pinges @macfarla Hey, I have resolved all the comments kindly review.

Thanks again!

**Comment by NickSneo** [comment]
@macfarla @pinges Hey, I have resolved all the comments. Can you please review again and merge?

**Review by macfarla — APPROVED** [review]
for me it looks good, @pinges can you confirm?

**Review by pinges — APPROVED** [review]
LGTM

**Pull request merged**